File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        res = set()

        def backtrack(i):
            max_size = len(res)
            if i == len(arr):
                return max_size

            for j in range(i, len(arr)):
                if len(set(arr[j])) == len(arr[j]) and not res & set(arr[j]):
                    res.update(arr[j])
                    max_size = max(backtrack(j + 1), max_size)
                    res.difference_update(arr[j])
            return max_size

        return backtrack(0)

File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    public int maxLength(List<String> arr) {
        Set<Character> uniques = new HashSet<>();
        return backtrack(0, arr, uniques);
    }

    private int backtrack(int idx, List<String> arr, Set<Character> uniques) {
        int maxSize = uniques.size();
        if (idx == arr.size()) {
            return maxSize;
        }
        // loop over all the elements
        for (int j = idx; j < arr.size(); j++) {
            if (isUnique(arr.get(j)) && isDisjoint(arr.get(j), uniques)) {
                for (char c : arr.get(j).toCharArray()) {   // update the uniques
                    uniques.add(c);
                }
                maxSize = Math.max(maxSize, backtrack(j + 1, arr, uniques));
                for (char c : arr.get(j).toCharArray()) {   // remove from uniques
                    uniques.remove(c);
                }
            }
        }
        return maxSize;
    }

    private boolean isUnique(String str) {
        Set<Character> set = new HashSet<>();
        for (char c : str.toCharArray()) {
            if (!set.add(c)) {
                return false;
            }
        }
        return true;
    }

    private boolean isDisjoint(String str, Set<Character> uniques) {
        for (char c : str.toCharArray()) {
            if (uniques.contains(c)) {      
                return false;
            }
        }
        return true;
    }
}